src/assets.zig

```diff
@@ -294,29 +310,64 @@ fn registerRecipesFromZon(zon: ZonElement) void {
 
 pub const Palette = struct { // MARK: Palette
 	palette: main.List([]const u8),
+
 	pub fn init(allocator: NeverFailingAllocator, zon: ZonElement, firstElement: ?[]const u8) !*Palette {
+		const self = switch(zon) {
+			.object => try loadFromZonLegacy(allocator, zon),
+			.array, .null => try loadFromZon(allocator, zon),
+			else => return error.InvalidPaletteFormat,
+		};
+
+		if(firstElement) |elem| {
+			if(self.palette.items.len == 0) {
+				self.palette.append(allocator.dupe(u8, elem));
+			}
+			std.debug.assert(std.mem.eql(u8, self.palette.items[0], elem));
+		}
+		return self;
+	}
+	fn loadFromZon(allocator: NeverFailingAllocator, zon: ZonElement) !*Palette {
+		const items = zon.toSlice();
+
 		const self = allocator.create(Palette);
 		self.* = Palette{
-			.palette = .init(allocator),
+			.palette = .initCapacity(allocator, @max(items.len, 128)),
 		};
 		errdefer self.deinit();
-		if(zon != .object or zon.object.count() == 0) {
-			if(firstElement) |elem| self.palette.append(allocator.dupe(u8, elem));
-		} else {
-			const palette = main.stackAllocator.alloc(?[]const u8, zon.object.count());
-			defer main.stackAllocator.free(palette);
-			for(palette) |*val| {
-				val.* = null;
-			}
-			var iterator = zon.object.iterator();
-			while(iterator.next()) |entry| {
-				palette[entry.value_ptr.as(usize, std.math.maxInt(usize))] = entry.key_ptr.*;
-			}
-			if(firstElement) |elem| std.debug.assert(std.mem.eql(u8, palette[0].?, elem));
-			for(palette) |val| {
-				std.log.info("palette[{}]: {s}", .{self.palette.items.len, val.?});
-				self.palette.append(allocator.dupe(u8, val orelse return error.MissingKeyInPalette));
+
+		for(items) |name| {
+			const stringId = name.as(?[]const u8, null) orelse return error.InvalidPaletteFormat;
+			self.palette.appendAssumeCapacity(allocator.dupe(u8, stringId));
+		}
+		return self;
+	}
+	fn loadFromZonLegacy(allocator: NeverFailingAllocator, zon: ZonElement) !*Palette {
+		// Using zon.object.count() here has the implication that array can not be sparse.
+		const paletteLength = zon.object.count();
+		const translationPalette = main.stackAllocator.alloc([]const u8, paletteLength);
```

Comment:
The `[]const u8` was deliberately null here, so we can catch cases where multiple entries map onto the same integer id, which could cause gaps in the palette too.